File: Android/bet/src/main/java/com/felipearpa/tyche/bet/pending/PendingBetItemView.kt
package com.felipearpa.tyche.bet.pending

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.material3.Button
import androidx.compose.material3.FilledTonalButton
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.max
import com.felipearpa.foundation.emptyString
import com.felipearpa.tyche.bet.PoolGamblerBetModel
import com.felipearpa.tyche.bet.awayTeamBetRawValue
import com.felipearpa.tyche.bet.homeTeamBetRawValue
import com.felipearpa.tyche.bet.poolGamblerBetDummyModel
import com.felipearpa.tyche.core.type.TeamScore
import com.felipearpa.tyche.ui.exception.UnknownLocalizedException
import com.felipearpa.tyche.ui.loading.BallSpinner
import com.felipearpa.tyche.ui.theme.LocalBoxSpacing
import com.felipearpa.ui.state.EditableViewState
import com.felipearpa.ui.state.relevantValue
import com.felipearpa.tyche.ui.R as SharedR

private val ICON_SIZE = 24.dp

@Composable
fun PendingBetItemView(viewModel: PendingBetItemViewModel, modifier: Modifier = Modifier) {
    val viewModelState by viewModel.state.collectAsState()
    var viewState by remember { mutableStateOf(PendingBetItemViewState.emptyVisualization()) }

    PendingBetItemView(
        viewModelState = viewModelState,
        viewState = viewState,
        onViewStateChanged = { newViewState -> viewState = newViewState },
        bet = {
            viewModel.bet(
                TeamScore(
                    homeTeamValue = viewState.value.homeTeamBet.toInt(),
                    awayTeamValue = viewState.value.awayTeamBet.toInt(),
                ),
            )
        },
        reset = viewModel::reset,
        retryBet = viewModel::retryBet,
        edit = { viewState = PendingBetItemViewState.Edition(viewState.value) },
        modifier = modifier,
    )

    LaunchedEffect(viewModelState) {
        viewState = when (val localViewModelState = viewModelState) {
            is EditableViewState.Initial -> {
                val poolGamblerBet = localViewModelState.value
                PendingBetItemViewState.Visualization(
                    PartialPoolGamblerBetModel(
                        homeTeamBet = poolGamblerBet.homeTeamBetRawValue(),
                        awayTeamBet = poolGamblerBet.awayTeamBetRawValue(),
                    ),
                )
            }

            else -> PendingBetItemViewState.Visualization(viewState.value)
        }
    }
}

@Composable
private fun PendingBetItemView(
    modifier: Modifier = Modifier,
    viewModelState: EditableViewState<PoolGamblerBetModel>,
    viewState: PendingBetItemViewState,
    onViewStateChanged: (PendingBetItemViewState) -> Unit = {},
    bet: () -> Unit = {},
    reset: () -> Unit = {},
    retryBet: () -> Unit = {},
    edit: () -> Unit = {},
) {
    Column(
        modifier = modifier,
        verticalArrangement = Arrangement.spacedBy(LocalBoxSpacing.current.medium),
    ) {
        when (viewModelState) {
            is EditableViewState.Initial, is EditableViewState.Success -> {
                val poolGamblerBet = when (viewModelState) {
                    is EditableViewState.Initial -> viewModelState.value
                    is EditableViewState.Success -> viewModelState.succeeded
                    else -> return
                }
                PendingBetItem(
                    poolGamblerBet = poolGamblerBet,
                    viewState = viewState,
                    onBetChanged = { newPartialPoolGamblerBet ->
                        onViewStateChanged(
                            viewState.copy(
                                newPartialPoolGamblerBet,
                            ),
                        )
                    },
                    modifier = Modifier.fillMaxWidth(),
                )
                DefaultActionBar(
                    viewModelState = viewModelState,
                    viewState = viewState,
                    bet = bet,
                    reset = reset,
                    edit = edit,
                    modifier = Modifier.fillMaxWidth(),
                )
            }

            is EditableViewState.Saving -> {
                PendingBetItem(
                    poolGamblerBet = viewModelState.target,
                    viewState = PendingBetItemViewState.Visualization(viewState.value),
                    modifier = Modifier.fillMaxWidth(),
                )
                LoadingActionBar(
                    viewModelState = viewModelState,
                    modifier = Modifier.fillMaxWidth(),
                )
            }

            is EditableViewState.Failure -> {
                PendingBetItem(
                    poolGamblerBet = viewModelState.failed,
                    viewState = PendingBetItemViewState.Visualization(viewState.value),
                    modifier = Modifier.fillMaxWidth(),
                )
                FailureActionBar(
                    viewModelState = viewModelState,
                    retryBet = retryBet,
                    reset = reset,
                    modifier = Modifier.fillMaxWidth(),
                )
            }
        }
    }
}

@Composable
private fun DefaultActionBar(
    viewModelState: EditableViewState<PoolGamblerBetModel>,
    viewState: PendingBetItemViewState,
    bet: () -> Unit,
    reset: () -> Unit,
    edit: () -> Unit,
    modifier: Modifier = Modifier,
) {
    if (viewState is PendingBetItemViewState.Edition) {
        EditableDefaultActionBar(
            viewModelState = viewModelState,
            viewState = viewState,
            bet = bet,
            reset = reset,
            modifier = modifier,
        )
    } else {
        NonEditableDefaultActionBar(
            viewModelState = viewModelState,
            viewState = viewState,
            edit = edit,
            modifier = modifier,
        )
    }
}

@Composable
private fun EditableDefaultActionBar(
    viewModelState: EditableViewState<PoolGamblerBetModel>,
    viewState: PendingBetItemViewState,
    bet: () -> Unit,
    reset: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val poolGamblerBet = viewModelState.relevantValue()
    var isRecomposition by remember { mutableStateOf(false) }
    var maxButtonWidth by remember { mutableStateOf(0.dp) }
    val density = LocalDensity.current
    val buttonWidthModifier = if (isRecomposition) Modifier.width(maxButtonWidth) else Modifier

    Row(
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically,
        modifier = modifier,
    ) {
        StateIndicator(viewModelState = viewModelState, isEditable = true)

        Row(horizontalArrangement = Arrangement.spacedBy(LocalBoxSpacing.current.medium)) {
            FilledTonalButton(
                onClick = reset,
                modifier = Modifier
                    .onSizeChanged { intSize ->
                        if (!isRecomposition) {
                            maxButtonWidth =
                                max(maxButtonWidth, with(density) { intSize.width.toDp() })
                        }
                    }
                    .then(buttonWidthModifier),
            ) {
                Text(text = stringResource(id = SharedR.string.cancel_action))
            }

            Button(
                onClick = bet,
                enabled = viewState.value != PartialPoolGamblerBetModel(
                    homeTeamBet = poolGamblerBet.homeTeamBetRawValue(),
                    awayTeamBet = poolGamblerBet.awayTeamBetRawValue(),
                ),
                modifier = Modifier
                    .onSizeChanged { intSize ->
                        if (!isRecomposition) {
                            maxButtonWidth =
                                max(maxButtonWidth, with(density) { intSize.width.toDp() })
                        }
                    }
                    .then(buttonWidthModifier),
            ) {
                Text(text = stringResource(id = SharedR.string.save_action))
            }
        }
    }

    LaunchedEffect(Unit) {
        isRecomposition = true
    }
}

@Composable
private fun NonEditableDefaultActionBar(
    viewModelState: EditableViewState<PoolGamblerBetModel>,
    viewState: PendingBetItemViewState,
    edit: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically,
        modifier = modifier,
    ) {
        StateIndicator(
            viewModelState = viewModelState,
            isEditable = viewState is PendingBetItemViewState.Edition,
        )

        TextButton(onClick = edit) {
            Text(text = stringResource(id = SharedR.string.edit_action))
        }
    }
}

@Composable
private fun LoadingActionBar(
    viewModelState: EditableViewState<PoolGamblerBetModel>,
    modifier: Modifier = Modifier,
) {
    Row(modifier = modifier) {
        StateIndicator(viewModelState = viewModelState)
    }
}

@Composable
private fun FailureActionBar(
    viewModelState: EditableViewState<PoolGamblerBetModel>,
    retryBet: () -> Unit,
    reset: () -> Unit,
    modifier: Modifier = Modifier,
) {
    var isRecomposition by remember { mutableStateOf(false) }
    var maxButtonWidth by remember { mutableStateOf(0.dp) }
    val density = LocalDensity.current
    val buttonWidthModifier = if (isRecomposition) Modifier.width(maxButtonWidth) else Modifier

    Row(
        modifier = modifier,
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        StateIndicator(viewModelState = viewModelState)

        Row(horizontalArrangement = Arrangement.spacedBy(LocalBoxSpacing.current.medium)) {
            FilledTonalButton(
                onClick = reset,
                modifier = Modifier
                    .onSizeChanged { intSize ->
                        if (!isRecomposition) {
                            maxButtonWidth =
                                max(maxButtonWidth, with(density) { intSize.width.toDp() })
                        }
                    }
                    .then(buttonWidthModifier),
            ) {
                Text(text = stringResource(id = SharedR.string.cancel_action))
            }

            Button(
                onClick = retryBet,
                modifier = Modifier
                    .onSizeChanged { intSize ->
                        if (!isRecomposition) {
                            maxButtonWidth =
                                max(maxButtonWidth, with(density) { intSize.width.toDp() })
                        }
                    }
                    .then(buttonWidthModifier),
            ) {
                Text(text = stringResource(id = SharedR.string.retry_action))
            }
        }
    }
    LaunchedEffect(Unit) {
        isRecomposition = true
    }
}

@Composable
private fun StateIndicator(
    viewModelState: EditableViewState<PoolGamblerBetModel>,
    isEditable: Boolean = false,
) {
    if (isEditable) {
        Icon(
            painter = painterResource(id = SharedR.drawable.pending),
            contentDescription = emptyString(),
        )
    } else {
        when (viewModelState) {
            is EditableViewState.Failure ->
                Icon(
                    painter = painterResource(id = SharedR.drawable.error),
                    contentDescription = emptyString(),
                    tint = MaterialTheme.colorScheme.error,
                )

            is EditableViewState.Saving -> BallSpinner(
                modifier = Modifier.size(ICON_SIZE),
            )

            is EditableViewState.Initial, is EditableViewState.Success -> {
                val poolGamblerBet = when (viewModelState) {
                    is EditableViewState.Initial -> viewModelState.value
                    is EditableViewState.Success -> viewModelState.succeeded
                    else -> return
                }
                ContentIndicator(poolGamblerBet = poolGamblerBet)
            }
        }
    }
}

@Composable
private fun ContentIndicator(poolGamblerBet: PoolGamblerBetModel) {
    if (poolGamblerBet.isLocked) {
        Icon(
            painter = painterResource(id = SharedR.drawable.pending),
            contentDescription = emptyString(),
        )
    } else {
        Icon(
            painter = painterResource(id = SharedR.drawable.done),
            contentDescription = emptyString(),
        )
    }
}

@Preview
@Composable
private fun NonEditableInitialPendingBetItemViewPreview() {
    MaterialTheme {
        Surface {
            PendingBetItemView(
                viewModelState = EditableViewState.Initial(poolGamblerBetDummyModel()),
                viewState = PendingBetItemViewState.Visualization(
                    partialPoolGamblerBetDummyModel(),
                ),
            )
        }
    }
}

@Preview
@Composable
private fun EditableInitialPendingBetItemViewPreview() {
    MaterialTheme {
        Surface {
            PendingBetItemView(
                viewModelState = EditableViewState.Initial(poolGamblerBetDummyModel()),
                viewState = PendingBetItemViewState.Edition(partialPoolGamblerBetDummyModel()),
            )
        }
    }
}

@Preview
@Composable
private fun NonEditableLoadingPendingBetItemViewPreview() {
    MaterialTheme {
        Surface {
            PendingBetItemView(
                viewModelState = EditableViewState.Saving(
                    current = poolGamblerBetDummyModel(),
                    target = poolGamblerBetDummyModel(),
                ),
                viewState = PendingBetItemViewState.Visualization(
                    partialPoolGamblerBetDummyModel(),
                ),
            )
        }
    }
}

@Preview
@Composable
private fun EditableLoadingPendingBetItemViewPreview() {
    MaterialTheme {
        Surface {
            PendingBetItemView(
                viewModelState = EditableViewState.Saving(
                    current = poolGamblerBetDummyModel(),
                    target = poolGamblerBetDummyModel(),
                ),
                viewState = PendingBetItemViewState.Edition(partialPoolGamblerBetDummyModel()),
            )
        }
    }
}

@Preview
@Composable
private fun NonEditableFailurePendingBetItemViewPreview() {
    MaterialTheme {
        Surface {
            PendingBetItemView(
                viewModelState = EditableViewState.Failure(
                    current = poolGamblerBetDummyModel(),
                    failed = poolGamblerBetDummyModel(),
                    exception = UnknownLocalizedException(),
                ),
                viewState = PendingBetItemViewState.Visualization(
                    partialPoolGamblerBetDummyModel(),
                ),
            )
        }
    }
}

@Preview
@Composable
private fun EditableFailurePendingBetItemViewPreview() {
    MaterialTheme {
        Surface {
            PendingBetItemView(
                viewModelState = EditableViewState.Failure(
                    current = poolGamblerBetDummyModel(),
                    failed = poolGamblerBetDummyModel(),
                    exception = UnknownLocalizedException(),
                ),
                viewState = PendingBetItemViewState.Edition(partialPoolGamblerBetDummyModel()),
            )
        }
    }
}

File: iOS/Bet/Sources/Bet/Pending/PendingBetItemView.swift
import SwiftUI
import UI
import Core
import ViewingState

struct PendingBetItemView: View {
    @StateObject private var viewModel: PendingBetItemViewModel
    @State private var viewState = PendingBetItemViewState.emptyVisualization()
    let poolGamblerBet: PoolGamblerBetModel

    init(
        viewModel: @autoclosure @escaping () -> PendingBetItemViewModel,
        poolGamblerBet: PoolGamblerBetModel
    ) {
        self._viewModel = .init(wrappedValue: viewModel())
        self.poolGamblerBet = poolGamblerBet
    }

    var body: some View {
        StatefulPendingBetItemView(
            viewModelState: viewModel.state ?? .idle(poolGamblerBet),
            viewState: $viewState,
            bet: {
                viewModel.bet(
                    betScore: TeamScore(
                        homeTeamValue: Int(viewState.value.homeTeamBet)!,
                        awayTeamValue: Int(viewState.value.awayTeamBet)!
                    )
                )
            },
            retryBet: viewModel.retryBet,
            reset: {
                viewState = .visualization(viewState.value)
                viewModel.reset()
            },
            edit: { viewState = .edition(viewState.value) }
        )
        .task(id: poolGamblerBet) { viewModel.bind(poolGamblerBet) }
        .onReceive(viewModel.$state) { state in
            guard let state = state else { return }
            let poolGamblerBet = state.activeValue()
            viewState = .visualization(poolGamblerBet.toPartialPoolGamblerBet())
        }
    }
}

private struct StatefulPendingBetItemView: View {
    let viewModelState: MutationState<PoolGamblerBetModel>
    @Binding var viewState: PendingBetItemViewState
    let bet: () -> Void
    let retryBet: () -> Void
    let reset: () -> Void
    let edit: () -> Void
    
    init(
        viewModelState: MutationState<PoolGamblerBetModel>,
        viewState: Binding<PendingBetItemViewState>,
        bet: @escaping () -> Void = {},
        retryBet: @escaping () -> Void = {},
        reset: @escaping () -> Void = {},
        edit: @escaping () -> Void = {}
    ) {
        self.viewModelState = viewModelState
        self._viewState = viewState
        self.bet = bet
        self.retryBet = retryBet
        self.reset = reset
        self.edit = edit
    }
    
    var body: some View {
        VStack {
            switch viewModelState {
            case .idle(let poolGamblerBet), .mutated(_, updated: let poolGamblerBet):
                PendingBetItem(
                    poolGamblerBet: poolGamblerBet,
                    viewState: $viewState
                )
                DefaultActionBar(
                    viewModelState: viewModelState,
                    viewState: $viewState,
                    bet: bet,
                    reset: reset,
                    edit: edit
                )
            case .mutating(_, updated: let poolGamblerBet):
                PendingBetItem(
                    poolGamblerBet: poolGamblerBet,
                    viewState: .constant(.visualization(poolGamblerBet.toPartialPoolGamblerBet()))
                )
                LoadingActionBar(viewModelState: viewModelState)
            case .failure(_, updated: let poolGamblerBet, _):
                PendingBetItem(
                    poolGamblerBet: poolGamblerBet,
                    viewState: .constant(.visualization(viewState.value))
                )
                FailureActionBar(
                    viewModelState: viewModelState,
                    retryBet: retryBet,
                    reset: reset
                )
            }
        }
    }
}

private struct DefaultActionBar: View {
    let viewModelState: MutationState<PoolGamblerBetModel>
    @Binding var viewState: PendingBetItemViewState
    let bet: () -> Void
    let reset: () -> Void
    let edit: () -> Void
    
    var body: some View {
        HStack(spacing: 8) {
            switch viewState {
            case .visualization:
                NonEditableDefaultActionBar(
                    viewModelState: viewModelState,
                    viewState: $viewState,
                    edit: edit
                )
            case .edition:
                EditableDefaultActionBar(
                    viewModelState: viewModelState,
                    viewState: $viewState,
                    bet: bet,
                    reset: reset
                )
            }
        }
    }
}

private struct EditableDefaultActionBar: View {
    let viewModelState: MutationState<PoolGamblerBetModel>
    @Binding var viewState: PendingBetItemViewState
    let bet: () -> Void
    let reset: () -> Void
    
    var body: some View {
        StateIndicator(state: viewModelState, isEditable: true)
        
        Spacer()
        
        Button(action: reset) {
            Text(sharedResource: .cancelAction)
        }.buttonStyle(.liquidGlass)

        Button(action: bet) {
            Text(sharedResource: .saveAction)
        }
        .buttonStyle(.liquidGlassProminent)
    }
}

private struct NonEditableDefaultActionBar: View {
    let viewModelState: MutationState<PoolGamblerBetModel>
    @Binding var viewState: PendingBetItemViewState
    let edit: () -> Void

    var body: some View {
        StateIndicator(state: viewModelState)

        Spacer()

        if !viewModelState.activeValue().isLocked {
            Button(action: edit) {
                Text(sharedResource: .editAction)
            }
        }
    }
}

private struct LoadingActionBar: View {
    let viewModelState: MutationState<PoolGamblerBetModel>

    var body: some View {
        HStack(spacing: 8) {
            StateIndicator(state: viewModelState)
            Spacer()
        }
    }
}

private struct FailureActionBar: View {
    let viewModelState: MutationState<PoolGamblerBetModel>
    let retryBet: () -> Void
    let reset: () -> Void
    
    var body: some View {
        HStack(spacing: 8) {
            StateIndicator(state: viewModelState)
            
            Spacer()
            
            Button(action: { reset() }) {
                Text(sharedResource: .cancelAction)
            }
            .buttonStyle(.liquidGlass)
            .tint(Color(sharedResource: .error))

            Button(action: { retryBet() }) {
                Text(sharedResource: .retryAction)
            }
            .buttonStyle(.liquidGlassProminent)
            .tint(Color(sharedResource: .error))
        }
    }
}

private struct StateIndicator: View {
    let viewModelstate: MutationState<PoolGamblerBetModel>
    let isEditable: Bool
    
    init(state: MutationState<PoolGamblerBetModel>, isEditable: Bool = false) {
        self.viewModelstate = state
        self.isEditable = isEditable
    }
    
    var body: some View {
        if isEditable {
            Image(sharedResource: .pending)
        } else {
            switch viewModelstate {
            case .failure:
                Image(sharedResource: .error)
                    .foregroundStyle(Color(sharedResource: .error))
            case .mutating:
                BallSpinner()
                    .frame(width: ICON_SIZE, height: ICON_SIZE)
            case .idle(let poolGamblerBet), .mutated(_, updated: let poolGamblerBet):
                ContentIndicator(poolGamblerBet: poolGamblerBet)
            }
        }
    }
}

private struct ContentIndicator: View {
    let poolGamblerBet: PoolGamblerBetModel
    
    var body: some View {
        if poolGamblerBet.isLocked {
            Image(sharedResource: .lock)
        } else {
            if poolGamblerBet.betScore == nil {
                Image(sharedResource: .pending)
            } else {
                Image(sharedResource: .done)
            }
        }
    }
}

private let ICON_SIZE: CGFloat = 24

#Preview("Non Editable Initial") {
    StatefulPendingBetItemView(
        viewModelState: .idle(poolGamblerBetDummyModel()),
        viewState: .constant(.visualization(partialPoolGamblerBetDummyModel()))
    )
}

#Preview("Editable Initial") {
    StatefulPendingBetItemView(
        viewModelState: .idle(poolGamblerBetDummyModel()),
        viewState: .constant(.edition(partialPoolGamblerBetDummyModel()))
    )
}

#Preview("Non Editable Loading") {
    StatefulPendingBetItemView(
        viewModelState: .mutating(
            original: poolGamblerBetDummyModel(),
            updated: poolGamblerBetDummyModel()
        ),
        viewState: .constant(.visualization(partialPoolGamblerBetDummyModel()))
    )
}

#Preview("Editable Loading") {
    StatefulPendingBetItemView(
        viewModelState: .mutating(
            original: poolGamblerBetDummyModel(),
            updated: poolGamblerBetDummyModel()
        ),
        viewState: .constant(.edition(partialPoolGamblerBetDummyModel()))
    )
}

#Preview("Non Editable Failure") {
    StatefulPendingBetItemView(
        viewModelState: .failure(
            original: poolGamblerBetDummyModel(),
            updated: poolGamblerBetDummyModel(),
            error: UnknownLocalizedError()
        ),
        viewState: .constant(.visualization(partialPoolGamblerBetDummyModel()))
    )
}

#Preview("Editable Failure") {
    StatefulPendingBetItemView(
        viewModelState: .failure(
            original: poolGamblerBetDummyModel(),
            updated: poolGamblerBetDummyModel(),
            error: UnknownLocalizedError()
        ),
        viewState: .constant(.edition(partialPoolGamblerBetDummyModel()))
    )
}
